config/cht/tasks.js

```diff
@@ -0,0 +1,528 @@
+        }
+      },
+      {
+        id: "pregnancy-home-visit",
+        start: 6,
+        end: 7,
+        dueDate: function (e, c, r) {
+          //console.log('r1', r);
+          return addDays(getMostRecentLMPDateForPregnancy(c, r), 38 * 7);
+        }
+      },
+      {
+        id: "pregnancy-home-visit",
+        start: 6,
+        end: 7,
+        dueDate: function (e, c, r) {
+          //console.log('r1', r);
+          return addDays(getMostRecentLMPDateForPregnancy(c, r), 40 * 7);
+        }
+      }
+    ]
+  },
+
+  //ANC Home Visit: show every 2 weeks (Unknown LMP)
+  {
+    icon: "icon-pregnancy",
+    title: "task.anc.pregnancy_home_visit.title",
+    appliesTo: "reports",
+    appliesToType: ["pregnancy"],
+    appliesIf: function (c, r) {// If LMP date is unknown
+      const recentLMP = getMostRecentLMPDateForPregnancy(c, r);
+      const x = recentLMP === null;
+      //We only want to show until 42 weeks + 7 days
+      const y = addDays(r.reported_date, 42 * 7 + 7) >= today;
+      //console.log('applies-f', x, y, r.reported_date, today);
+      return x && y;
+    },
+
+    resolvedIf: function (c, r, event, dueDate) {
+      //console.log('ddr', dueDate);
+      if (r.reported_date < getNewestDeliveryTimestamp(c)) return true;//delivery form submitted
+      if (r.reported_date < getNewestPregnancyTimestamp(c)) return true;//old pregnancy report
+
+      //miscarriage or abortion
+      if (getRecentANCVisitWithEvent(c, r, 'abortion') || getRecentANCVisitWithEvent(c, r, 'miscarriage')) return true;
+
+      //(refused or migrated) and clear tasks 
+      if (isPregnancyTaskMuted(c)) return true;
+
+      let x = isFormArraySubmittedInWindow(c.reports, ['pregnancy_home_visit'],
+        Math.max(addDays(dueDate, -event.start).getTime(), r.reported_date),
+        addDays(dueDate, event.end + 1).getTime());
+      //console.log('resolved by visit report', x);
+      //let x = false;
+      //console.log('resolved', x);
+      return x;
+    },
+
+    actions: [
+      {
+        type: "report",
+        form: "pregnancy_home_visit",
+        label: "Pregnancy home visit",
+        modifyContent: function (content, c, r) {
+          content.source = 'task';
+          content.source_id = r._id;
+          content.contact = c.contact;
+          content.pregnancy_uuid = r._id;
+          //content.lmp_date_8601 = r.fields.lmp_date_8601;
+          //content.edd_8601 = r.fields.edd_8601;
+          //content.weeks_since_lmp = r.fields.weeks_since_lmp;
+        }
+      }
+    ],
+    events: [
+      {
+        id: "pregnancy-home-visit",
+        start: 6,
+        end: 7,
+        dueDate: function (e, c, r) { //every two weeks since registration
+          //console.log('now', now);
+          let upcomingDate = addDays(getDateMS(r.reported_date), 14);
+          let countLoops = 0;
+          while ((upcomingDate < addDays(r.reported_date, 14) || upcomingDate < addDays(today, -7))) {
```

Comment:
I explained a method in another comment about using a function which returns an event object in order to make this sort of multi-event code a bit more elegant. Would that be more suitable here than this complex iteration?

Personally, I find this code hard to read. If you keep it, I think a comment explaining the criteria in plain english would be helpful.

Reply:
It is now changed, the reusable event generating function is now used. Thanks.